### PR TITLE
Ofp4 sliced fields

### PR DESCRIPTION
**Pull request by blp**
This series fixes a bunch of small bugs I noticed, then the final commit makes VLAN matches much closer to working.

### CODE REVIEW

File: ofp4/backend.cpp [hunk 1]
```diff
@@ -134,16 +134,43 @@ class ActionTranslator : public Inspector {
...
 
-                name = field->externalName();
-                auto size = field->type->width_bits();
-                currentTranslation = new IR::OF_Register(name, size, 0, size - 1);
+                // This field might be a slice of an OpenFlow field or it might
+                // be the whole field.
+                int size, low, high;
+                if (auto slice = field->getAnnotation("of_slice")) {
+                    if (slice->expr.size() != 3) {
+                        ::error(ErrorType::ERR_UNKNOWN, "%1%: @of_slice must contain 3 constants", slice);
```
[the remark this answers is not in the document]
Comment (blp) [reply ofp4/backend.cpp]: Thanks, I pushed a fix.

File: ofp4/backend.cpp [hunk 2]
```diff
@@ -134,16 +134,43 @@ class ActionTranslator : public Inspector {
...
+                        *x = slice->expr[i++]->checkedTo<IR::Constant>()->asInt();
+                    }
+                    if (!(0 <= low && low <= high && high < size)) {
+                        ::error(ErrorType::ERR_UNKNOWN, "%1%: @of_slice(low,high,size) requires 0 <= low <= high < size", slice);
+                        return false;
+                    }
+
+                    int width = field->type->width_bits();
+                    if (high - low + 1 != width) {
+                        ::error(ErrorType::ERR_UNKNOWN, "%1%: @of_slice(low,high,size) is a %1%-bit slice but %2% is a %3%-bit field.",
```
[the remark this answers is not in the document]
Comment (blp) [reply ofp4/backend.cpp]: Thanks, I pushed a fix.

File: ofp4/backend.cpp [hunk 3]
```diff
@@ -134,16 +134,43 @@ class ActionTranslator : public Inspector {
...
+                // be the whole field.
+                int size, low, high;
+                if (auto slice = field->getAnnotation("of_slice")) {
+                    if (slice->expr.size() != 3) {
+                        ::error(ErrorType::ERR_UNKNOWN, "%1%: @of_slice must contain 3 constants", slice);
+                        return false;
+                    }
+                    int i = 0;
+                    for (auto x : { &low, &high, &size }) {
+                        *x = slice->expr[i++]->checkedTo<IR::Constant>()->asInt();
```
[the remark this answers is not in the document]
Comment (blp) [reply ofp4/backend.cpp:153]: Doesn't PARSE_CONSTANT_LIST in ParseAnnotations (in midend.cpp) ensure that the expr is only Constants?

[the remark this answers is not in the document]
Comment (mbudiu-vmw) [reply ofp4/backend.cpp:153]: Defensive programming can't hurt. Some pass may have modified the annotations, for example.

[the remark this answers is not in the document]
Comment (mbudiu-vmw) [reply ofp4/backend.cpp:153]: But yes, it seems like in this case this is correct, but I actually don't know what error will be reported if they aren't constant.

**Review by mbudiu-vmw — APPROVED** [review]

File: ofp4/backend.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (blp) [reply ofp4/backend.cpp]: OK, no big deal, I pushed a defensive version.

File: ofp4/backend.cpp [hunk 4]
```diff
@@ -150,7 +150,13 @@ class ActionTranslator : public Inspector {
                     }
                     int i = 0;
                     for (auto x : { &low, &high, &size }) {
-                        *x = slice->expr[i++]->checkedTo<IR::Constant>()->asInt();
+                        auto elem = slice->expr[i++];
+                        auto value = elem->checkedTo<IR::Constant>();
```
Comment (mbudiu-vmw) [inline ofp4/backend.cpp:154]: this has to be just `->to`.

Comment (blp) [reply ofp4/backend.cpp:154]: Oops, I'm dumb. I fixed it.

**Pull request merged**